File: apps/web/types/release-ops.ts
// FM-131–137: Release Operations types

export interface ReleasePackage {
  id: string;
  project_id: string;
  run_id: string;
  version: string;
  status: string;
  summary: string | null;
  artifact_manifest: Record<string, unknown> | null;
  changelog: Record<string, unknown> | null;
  confidence_snapshot: Record<string, unknown> | null;
  rollback_metadata: Record<string, unknown> | null;
  target_environment_id: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface ReleasePackageList {
  items: ReleasePackage[];
  total: number;
}

export interface DeploymentEnvironment {
  id: string;
  project_id: string;
  name: string;
  tier: string;
  description: string | null;
  config: Record<string, unknown> | null;
  required_gates: Record<string, unknown> | null;
  promotion_target_id: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface EnvironmentList {
  items: DeploymentEnvironment[];
  total: number;
}

export interface GateResult {
  id: string;
  release_package_id: string;
  gate_name: string;
  gate_status: string;
  detail: string | null;
  metadata_: Record<string, unknown> | null;
  evaluated_at: string;
}

export interface GateResultList {
  items: GateResult[];
  total: number;
}

export interface ReadinessCheck {
  check: string;
  passed: boolean;
  detail: string;
  score?: number;
}

export interface ReadinessReport {
  release_package_id: string;
  environment_id: string;
  is_ready: boolean;
  checks: ReadinessCheck[];
  blockers: string[];
  confidence_score: number;
  passed_checks: number;
  total_checks: number;
}

export interface RollbackReadiness {
  release_package_id: string;
  is_rollback_ready: boolean;
  recovery_points: Array<{
    type: string;
    id: string;
    label: string;
    [key: string]: unknown;
  }>;
  recovery_point_count: number;
  strategies: Array<{
    strategy: string;
    description: string;
    available: string;
  }>;
  risk_signals: Array<{
    signal: string;
    level: string;
    detail: string;
  }>;
  risk_level: string;
}

export interface GateEvaluation {
  release_package_id: string;
  total_gates: number;
  passed: number;
  failed: number;
  all_passed: boolean;
  gate_results: Array<{
    gate: string;
    status: string;
    detail: string;
  }>;
  package_status: string;
}

export interface TimelineEntry {
  timestamp: string | null;
  category: string;
  event: string;
  detail: string;
}

export interface OperationalTimeline {
  run_id: string;
  project_id: string;
  run_status: string;
  total_entries: number;
  categories: Record<string, number>;
  timeline: TimelineEntry[];
}

File: apps/api/app/schemas/release_ops.py
"""FM-131–137: Release operations Pydantic schemas."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.release_ops import (
    EnvironmentTier,
    GateStatus,
    ReleaseStatus,
)


# ---------------------------------------------------------------------------
# FM-131: Release Package schemas
# ---------------------------------------------------------------------------


class ReleasePackageCreate(BaseModel):
    version: str = Field(..., min_length=1, max_length=100)
    summary: str = Field(..., min_length=1)
    target_environment_id: uuid.UUID | None = None
    artifact_manifest: dict | None = None
    changelog: dict | None = None
    created_by: str | None = Field(None, max_length=100)


class ReleasePackageUpdate(BaseModel):
    version: str | None = Field(None, max_length=100)
    summary: str | None = None
    status: ReleaseStatus | None = None
    target_environment_id: uuid.UUID | None = None
    artifact_manifest: dict | None = None
    changelog: dict | None = None
    rollback_metadata: dict | None = None


class ReleasePackageRead(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    run_id: uuid.UUID
    version: str
    status: ReleaseStatus
    summary: str
    artifact_manifest: dict | None
    changelog: dict | None
    confidence_snapshot: dict | None
    rollback_metadata: dict | None
    target_environment_id: uuid.UUID | None
    created_by: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ReleasePackageList(BaseModel):
    items: list[ReleasePackageRead]
    total: int


# ---------------------------------------------------------------------------
# FM-132: Deployment Environment schemas
# ---------------------------------------------------------------------------


class EnvironmentCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    tier: EnvironmentTier = EnvironmentTier.DEVELOPMENT
    description: str | None = None
    config: dict | None = None
    required_gates: dict | None = None
    promotion_target_id: uuid.UUID | None = None


class EnvironmentUpdate(BaseModel):
    name: str | None = Field(None, max_length=200)
    tier: EnvironmentTier | None = None
    description: str | None = None
    config: dict | None = None
    required_gates: dict | None = None
    promotion_target_id: uuid.UUID | None = None
    is_active: bool | None = None


class EnvironmentRead(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    name: str
    tier: EnvironmentTier
    description: str | None
    config: dict | None
    required_gates: dict | None
    promotion_target_id: uuid.UUID | None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class EnvironmentList(BaseModel):
    items: list[EnvironmentRead]
    total: int


# ---------------------------------------------------------------------------
# FM-134: Release Gate Result schemas
# ---------------------------------------------------------------------------


class GateResultRead(BaseModel):
    id: uuid.UUID
    release_package_id: uuid.UUID
    gate_name: str
    gate_status: GateStatus
    detail: str | None
    metadata_: dict | None = Field(None, alias="metadata_")
    evaluated_at: datetime

    model_config = {"from_attributes": True, "populate_by_name": True}


class GateResultList(BaseModel):
    items: list[GateResultRead]
    total: int
